Time based sharding tutorial (#1805)
qdrant/landing_page: qdrant-landing/content/documentation/tutorials-operations/time-based-sharding.md

---
title: Time-Based Sharding
weight: 35
---

# Time-Based Sharding in Qdrant

> When working with massive, fast-moving datasets, like social media streams, efficient storage and retrieval are critical. Often, only the most recent data is relevant, while older data can be archived or deleted. For instance, in sentiment analysis of social media posts, you might only need the last 7 days of data to capture current trends, with most queries focusing on the last 24 hours.

Reviewer: ```suggestion
When working with massive, fast-moving datasets, like social media or image/video streams, efficient storage and retrieval are critical. Often, only the most recent data is relevant, while older data can be archived or deleted. For instance, in sentiment analysis of social media posts, you might only need the last 7 days of data to capture current trends, with most queries focusing on the last 24 hours.
```

> Storing everything in a regular Qdrant collection can lead to expensive re-indexing across the entire dataset when deleting old points, impacting performance. A better solution is **time-based sharding**, where data is routed to a specific [shard](/documentation/guides/distributed_deployment/?q=shard#sharding) based on its timestamp. Segmenting data by age enables easy data retention and efficient querying of recent data.

Reviewer: ```suggestion
Storing everything in Qdrant collection with default sharding can lead to expensive re-indexing across the entire dataset when deleting old points, impacting performance. A better solution is **time-based sharding**, where points are routed to a specific [shard (or shards)](/documentation/guides/distributed_deployment/#sharding) based on timestamp. For use cases with a natural time-to-live (TTL) segmentation, sharding by a timestamp-based key enables efficient querying of recent data and allows users to seamlessly drop the old.
```

For example, with daily shards, today's data is stored in today's shard, yesterday's data in yesterday's shard, and so on. Queries can target specific shards (today's shard, for example) or multiple shards to cover a date range.

<figure>
  <img src="/documentation/tutorials/time-based-sharding/time-based-sharding.png">
  <figcaption>
    Time-based sharding routes data to different shards based on timestamp. Typically, all writes go to the newest shard, while queries can target one or more shards. Older shards can be pruned in the background without affecting performance.
  </figcaption>
</figure>

> Note that daily sharding is just one example. Depending on your data volume and retention needs, you could shard by hour, week, month, or any other time interval that suits your use case.

Reviewer: ```suggestion
Depending on your data volume and retention needs, you could shard by hour, week, month, or any other time interval that suits your use case.
```

This tutorial assumes you are using [Qdrant Cloud Inference](/documentation/concepts/inference/#qdrant-cloud-inference) to generate vector embeddings. If you manage your own embedding infrastructure, you can apply the same principles, but you will need to adapt the code examples to use your embedding service.

## Create Collection

Create a collection with [user-defined sharding](/documentation/guides/distributed_deployment/#user-defined-sharding) by setting the sharding method to custom.

{{< code-snippet path="/documentation/headless/snippets/time-based-sharding/" block="create-collection" >}}

Custom shards can be accessed by their shard key. In this tutorial, the shard keys are the dates in `YYYY-MM-DD` format, extracted from the timestamp of each data point.

For default collections using `auto` sharding, `shard_number` determines the total number of shards for a collection. With user-defined sharding, it determines the number of shards **per shard key**: the total number of shards for a collection equals the number of shard keys (days) multiplied by the `shard_number`.

> Setting `shard_number` to `1` creates a single shard for each shard key (a separate shard for each day of data). For very large datasets, you can improve write throughput by increasing `shard_number` and distributing these shards across multiple peers in the cluster. However, avoid creating too many shards, as each shard consumes resources, and having too many shards can lead to performance degradation.

Reviewer: > However, avoid creating too many shards, as each shard consumes resources, and having too many shards can lead to performance degradation.

Why?

Reviewer: How should users determine how to set `shard_number` when leveraging time-based sharding? What is the default value?

Reviewer 2: I have reworded this paragraph to make it clear that the default `shard_number` is 1. W.r. to determining what to set the number to, I've added  _Test what the optimal number of shards is for your dataset and cluster configuration._

## Ingest Historical Data

Time series data often arrives in streams, with new data points continuously being added. Each data point may include a timestamp indicating when it was created. You can use this timestamp to determine which shard a data point belongs to. If your data does not have timestamps, you can use the current time.

This tutorial uses a [sample dataset of social media posts](https://github.com/qdrant/examples/blob/master/time-based-sharding/social-media-posts.csv) with timestamps. Let's assume today is April 7th, 2026. You'll start by ingesting some historical data from this week (April 1-7). Here are a few sample rows from the dataset:

| datetime            | text
|---|---|
| 2026-04-06T09:04:28 | April sunshine through the office window makes everything better.
| 2026-04-06T09:04:32 | Morning stretch, good coffee, clear intentions. Monday: sorted.
| 2026-04-06T09:05:52 | Grateful for a productive first day of the week.

Upload the dataset and store each day of data in its own shard:

{{< code-snippet path="/documentation/headless/snippets/time-based-sharding/" block="upload-vectors" >}}

Let's break down the code:

> - First, a list of existing shard keys in the collection is retrieved. There should be none because you just created the collection, but in production, this ensures you take into account any existing data.

Reviewer: Are there any consequences of trying to execute a patch command to a collection, after creating it with custom sharding, but before creating the first shard key?

Reviewer 2: What kind of patch command?

Reviewer: For example, changing the HNSW parameters?

Reviewer: We discussed this on a call and tested things out, creating collection with custom sharding. We were able to patch the collection and create payload indexes before creating shard keys
